### ISSUE TITLE
Crashing during export to mesh + textures

### ISSUE THREAD

**Issue by Lhun** [introlab/rtabmap#136]
Hi there again. I have an extremely capable system, and I'm getting a full crash when attempting to export a scan to mesh. 

It could be that it's too complex, I'm not sure. I'm looking to do simple room scans for use in visualization, and I'm having no luck so far. 

Using a Kinect360, what settings should I be using for visualization? The mapping is very rapid and impressive, but I'm not sure how to take this 5m+ point cloud and get it out of the software.

**Comment by matlabbe** [comment]
Which version (on which system) are you using? You can start RTAB-Map from the console/terminal and set logger level to Debug (Preferences->Logging->Level). Try exporting the mesh and copy/part the terminal output here.

**Comment by matlabbe** [comment]
A print screen of your Export dialog could also help to reproduce the problem.

**Comment by Lhun** [comment]
Win64.
Thanks for the swift response. Creating a log right now.

**Comment by Lhun** [comment]
okay, so nothing exciting, it seems to crash exporting the temp textures: 
```

[ INFO] (2016-10-30 17:40:38) util3d_surface.cpp:530::rtabmap::util3d::createTextureCameras() Saved temporary texture: "C:/Users/Lhun/Documents/RTAB-Map/tmp_textures/texture_282.png"

[ INFO] (2016-10-30 17:40:38) util3d_surface.cpp:530::rtabmap::util3d::createTextureCameras() Saved temporary texture: "C:/Users/Lhun/Documents/RTAB-Map/tmp_textures/texture_283.png"

[ INFO] (2016-10-30 17:40:38) util3d_surface.cpp:530::rtabmap::util3d::createTextureCameras() Saved temporary texture: "C:/Users/Lhun/Documents/RTAB-Map/tmp_textures/texture_284.png"

[ INFO] (2016-10-30 17:40:38) util3d_surface.cpp:530::rtabmap::util3d::createTextureCameras() Saved temporary texture: "C:/Users/Lhun/Documents/RTAB-Map/tmp_textures/texture_285.png"

[ INFO] (2016-10-30 17:40:38) util3d_surface.cpp:530::rtabmap::util3d::createTextureCameras() Saved temporary texture: "C:/Users/Lhun/Documents/RTAB-Map/tmp_textures/texture_28

```
![export1](https://cloud.githubusercontent.com/assets/23122906/19840314/544c4e7a-9ec9-11e6-82ba-53df7324ac76.png)
![export2](https://cloud.githubusercontent.com/assets/23122906/19840317/58587034-9ec9-11e6-9ab5-ac799aab024e.png)

**Comment by Lhun** [comment]
Just trying to export a reasonable mesh - if anyone has any settings config saved for kinect360 that is known to work reasonably well with a 2-5 pass scan on a 4m square space to export as obj+mtl, I would be forever in your debt. Looking to get a workflow going for VR recreation and mapping.

**Comment by matlabbe** [comment]
I cannot reproduce the crash, I'll try on Windows tomorrow.

You can also try the "Organized Cloud" reconstruction flavor, which gives a lot better looking textured meshes (note that there is an [issue](https://github.com/introlab/rtabmap/issues/137) on dense texture mesh exportation). With organized clouds, I recommend to use "Min polygon cluster size" instead of "Noise filtering" to remove noisy polygons. These parameters would give you nice results:
![untitled](https://cloud.githubusercontent.com/assets/2319645/19840664/871d4586-9ed1-11e6-8202-1f44b89f0a1b.jpg)

To reduce the number of polygons, use "3D cloud decimation" under "Regenerate clouds". You can also activate Preferences->General Settings->Cloud Filtering->Pose Filtering to avoid creating meshes for supperposed nodes.

**Comment by Lhun** [comment]
That's fantastic, thank you, I'll give that a shot. Would you recommend orb
over grift in this case, for a purely visual aspect?

**Comment by Lhun** [comment]
So I went ahead and did this, however, the output, rather then being an obj, ended up being several hundred ply files, and they were not assembled, remained as point clouds, and had no references to texture. Maybe I'm missing a step, apologies for the wet behind the ears here, but I assumed the output would be a single mesh+texture.

**Comment by matlabbe** [comment]
I tried your settings on Windows (latest release 0.11.11), and I cannot reproduce the crash. Can you share your database?

For the organized cloud setting, yes by checking "Assemble clouds/meshes", only one mesh should be saved. If "Texture mapping" is checked, then an OBJ is generated instead of PLY. I tried my settings above and I get this OBJ (opened in MeshLab):
![capture](https://cloud.githubusercontent.com/assets/2319645/19865099/3687ab0c-9f71-11e6-802e-9929688bd1a6.PNG)

Can you show the output of the progress dialog? Mine is:
```
17:45:47 Copied cloud 1 from cache with 19200 points and 15182 indices (1/32).
17:45:47 Copied cloud 3 from cache with 19200 points and 15717 indices (2/32).
17:45:47 Copied cloud 4 from cache with 19200 points and 15542 indices (3/32).
17:45:47 Copied cloud 5 from cache with 19200 points and 15345 indices (4/32).
17:45:47 Copied cloud 6 from cache with 19200 points and 14094 indices (5/32).
17:45:47 Copied cloud 7 from cache with 19200 points and 13076 indices (6/32).
17:45:47 Copied cloud 8 from cache with 19200 points and 11870 indices (7/32).
17:45:47 Copied cloud 9 from cache with 19200 points and 14925 indices (8/32).
17:45:47 Copied cloud 10 from cache with 19200 points and 15451 indices (9/32).
17:45:47 Copied cloud 11 from cache with 19200 points and 15846 indices (10/32).
17:45:47 Copied cloud 12 from cache with 19200 points and 15785 indices (11/32).
17:45:47 Copied cloud 13 from cache with 19200 points and 16506 indices (12/32).
17:45:48 Copied cloud 14 from cache with 19200 points and 15970 indices (13/32).
17:45:48 Copied cloud 15 from cache with 19200 points and 15776 indices (14/32).
17:45:48 Copied cloud 16 from cache with 19200 points and 14345 indices (15/32).
17:45:48 Copied cloud 17 from cache with 19200 points and 11457 indices (16/32).
17:45:48 Copied cloud 18 from cache with 19200 points and 7987 indices (17/32).
17:45:48 Copied cloud 19 from cache with 19200 points and 9320 indices (18/32).
17:45:48 Copied cloud 20 from cache with 19200 points and 12137 indices (19/32).
17:45:48 Copied cloud 21 from cache with 19200 points and 13497 indices (20/32).
17:45:48 Copied cloud 22 from cache with 19200 points and 15132 indices (21/32).
17:45:48 Copied cloud 23 from cache with 19200 points and 15062 indices (22/32).
17:45:48 Copied cloud 24 from cache with 19200 points and 14431 indices (23/32).
17:45:48 Copied cloud 26 from cache with 19200 points and 12459 indices (24/32).
17:45:48 Copied cloud 27 from cache with 19200 points and 11668 indices (25/32).
17:45:48 Copied cloud 28 from cache with 19200 points and 13602 indices (26/32).
17:45:48 Copied cloud 29 from cache with 19200 points and 14394 indices (27/32).
17:45:48 Copied cloud 30 from cache with 19200 points and 15458 indices (28/32).
17:45:48 Copied cloud 31 from cache with 19200 points and 15736 indices (29/32).
17:45:48 Copied cloud 32 from cache with 19200 points and 15616 indices (30/32).
17:45:48 Copied cloud 33 from cache with 19200 points and 16489 indices (31/32).
17:45:48 Copied cloud 34 from cache with 19200 points and 16298 indices (32/32).
17:45:48 Gain compensation of 32 clouds...
17:45:49 Applying gain compensation...
17:45:49 Cloud 1 has gain 0.861344
17:45:49 Cloud 3 has gain 0.858267
17:45:49 Cloud 4 has gain 0.853592
17:45:49 Cloud 5 has gain 1.09896
17:45:49 Cloud 6 has gain 0.649708
17:45:49 Cloud 7 has gain 0.601025
17:45:49 Cloud 8 has gain 0.653189
17:45:49 Cloud 9 has gain 1.01786
17:45:49 Cloud 10 has gain 1.33935
17:45:49 Cloud 11 has gain 0.92196
17:45:49 Cloud 12 has gain 0.911898
17:45:49 Cloud 13 has gain 1.04661
17:45:49 Cloud 14 has gain 0.936551
17:45:49 Cloud 15 has gain 0.896505
17:45:49 Cloud 16 has gain 0.826169
17:45:49 Cloud 17 has gain 1.50536
17:45:49 Cloud 18 has gain 1.29203
17:45:49 Cloud 19 has gain 1.38673
17:45:49 Cloud 20 has gain 1.32677
17:45:49 Cloud 21 has gain 1.06434
17:45:49 Cloud 22 has gain 0.729587
17:45:49 Cloud 23 has gain 0.646278
17:45:49 Cloud 24 has gain 0.642561
17:45:49 Cloud 26 has gain 0.636722
17:45:49 Cloud 27 has gain 0.64603
17:45:49 Cloud 28 has gain 0.634262
17:45:49 Cloud 29 has gain 0.976196
17:45:49 Cloud 30 has gain 0.952556
17:45:49 Cloud 31 has gain 0.879196
17:45:49 Cloud 32 has gain 0.787039
17:45:49 Cloud 33 has gain 0.820833
17:45:49 Cloud 34 has gain 0.817773
17:45:49 Organized fast mesh... 
17:45:49 Mesh 1 created with 27309 polygons (1/32).
17:45:49 Mesh 3 created with 28409 polygons (2/32).
17:45:49 Mesh 4 created with 27959 polygons (3/32).
17:45:49 Mesh 5 created with 27560 polygons (4/32).
17:45:49 Mesh 6 created with 23314 polygons (5/32).
17:45:49 Mesh 7 created with 21784 polygons (6/32).
17:45:49 Mesh 8 created with 20958 polygons (7/32).
17:45:49 Mesh 9 created with 27170 polygons (8/32).
17:45:49 Mesh 10 created with 28003 polygons (9/32).
17:45:49 Mesh 11 created with 28726 polygons (10/32).
17:45:49 Mesh 12 created with 28408 polygons (11/32).
17:45:49 Mesh 13 created with 29065 polygons (12/32).
17:45:49 Mesh 14 created with 26927 polygons (13/32).
17:45:49 Mesh 15 created with 28986 polygons (14/32).
17:45:49 Mesh 16 created with 26220 polygons (15/32).
17:45:49 Mesh 17 created with 20629 polygons (16/32).
17:45:49 Mesh 18 created with 14678 polygons (17/32).
17:45:49 Mesh 19 created with 17418 polygons (18/32).
17:45:49 Mesh 20 created with 22467 polygons (19/32).
17:45:49 Mesh 21 created with 24489 polygons (20/32).
17:45:49 Mesh 22 created with 27499 polygons (21/32).
17:45:49 Mesh 23 created with 26588 polygons (22/32).
17:45:49 Mesh 24 created with 25201 polygons (23/32).
17:45:49 Mesh 26 created with 20801 polygons (24/32).
17:45:49 Mesh 27 created with 18897 polygons (25/32).
17:45:49 Mesh 28 created with 23487 polygons (26/32).
17:45:49 Mesh 29 created with 25302 polygons (27/32).
17:45:49 Mesh 30 created with 28490 polygons (28/32).
17:45:49 Mesh 31 created with 29276 polygons (29/32).
17:45:49 Mesh 32 created with 28391 polygons (30/32).
17:45:49 Mesh 33 created with 30641 polygons (31/32).
17:45:49 Mesh 34 created with 29398 polygons (32/32).
17:45:49 TextureMesh 1 created [cameras=1] (1/32).
17:45:49 TextureMesh 3 created [cameras=1] (2/32).
17:45:49 TextureMesh 4 created [cameras=1] (3/32).
17:45:49 TextureMesh 5 created [cameras=1] (4/32).
17:45:50 TextureMesh 6 created [cameras=1] (5/32).
17:45:50 TextureMesh 7 created [cameras=1] (6/32).
17:45:50 TextureMesh 8 created [cameras=1] (7/32).
17:45:50 TextureMesh 9 created [cameras=1] (8/32).
17:45:50 TextureMesh 10 created [cameras=1] (9/32).
17:45:50 TextureMesh 11 created [cameras=1] (10/32).
17:45:50 TextureMesh 12 created [cameras=1] (11/32).
17:45:50 TextureMesh 13 created [cameras=1] (12/32).
17:45:50 TextureMesh 14 created [cameras=1] (13/32).
17:45:50 TextureMesh 15 created [cameras=1] (14/32).
17:45:50 TextureMesh 16 created [cameras=1] (15/32).
17:45:50 TextureMesh 17 created [cameras=1] (16/32).
17:45:50 TextureMesh 18 created [cameras=1] (17/32).
17:45:50 TextureMesh 19 created [cameras=1] (18/32).
17:45:50 TextureMesh 20 created [cameras=1] (19/32).
17:45:50 TextureMesh 21 created [cameras=1] (20/32).
17:45:50 TextureMesh 22 created [cameras=1] (21/32).
17:45:50 TextureMesh 23 created [cameras=1] (22/32).
17:45:50 TextureMesh 24 created [cameras=1] (23/32).
17:45:50 TextureMesh 26 created [cameras=1] (24/32).
17:45:50 TextureMesh 27 created [cameras=1] (25/32).
17:45:50 TextureMesh 28 created [cameras=1] (26/32).
17:45:50 TextureMesh 29 created [cameras=1] (27/32).
17:45:50 TextureMesh 30 created [cameras=1] (28/32).
17:45:50 TextureMesh 31 created [cameras=1] (29/32).
17:45:50 TextureMesh 32 created [cameras=1] (30/32).
17:45:50 TextureMesh 33 created [cameras=1] (31/32).
17:45:50 TextureMesh 34 created [cameras=1] (32/32).
17:45:50 Assembling 32 meshes...
17:45:54 Saving the mesh (with 32 textures)...
17:46:22 Saving the mesh (with 32 textures)... done.
```

**Comment by Lhun** [comment]
I don't have access to that system at the moment, but IIRC it skips the
"assembling" portion at the end.

**Comment by Lhun** [comment]
If you would like, I have another instance running here on alternate hardware (also 11.11 win64) I could try the database you just meshed with the same settings and see if I'm successful,

**Comment by matlabbe** [comment]
I didn't keep the previous database, so I just created a new one: [mesh_test.db.zip](https://github.com/introlab/rtabmap/files/562699/mesh_test.db.zip)

Click on this image to see the OBJ in Sketchfab
[![screenshot from 2016-10-31 15-25-37](https://cloud.githubusercontent.com/assets/2319645/19868814/1de81a6e-9f80-11e6-9487-87d2be47b409.png)](https://skfb.ly/VvCD)

**Comment by Lhun** [comment]
okay wow, that's a trip--- It's exported perfectly on my machine here at
work, as an obj. Thanks so much - I must have changed something in the
settings or the scan is simply too big? I'm not sure. I'll try another
small export, and I'll reset all the settings.

I did notice that when I imported the db it had several settings different
then the defaults (I believe it said 4) so what I'll do is globally reset,
import your db, save the settings, then try an export over there. Cheers!
I'll let you know how it goes at home.

**Comment by Lhun** [comment]
[image: Inline image 1]

Go figure. Happy halloween, and thanks again. Great piece of software you
have here.

**Comment by Lhun** [comment]
Okay, so still getting issues at home. I have remote access today. 

The log did indeed show something different this time. 
Using your db, I get the following: 
![image](https://cloud.githubusercontent.com/assets/23122906/19897381/3dc748ba-a02e-11e6-9698-219502fbea2e.png)

Which then prompts to export as a ply rather then an obj. It's quite strange and I'm thinking I'll do a wipe and re-install just to be sure. 

First, the only difference between a master reset and your DB settings is: 
![image](https://cloud.githubusercontent.com/assets/23122906/19897295/ea897a56-a02d-11e6-8c97-6e1c9f3fdf83.png)

I also got this in the log after attempting an export: 

![image](https://cloud.githubusercontent.com/assets/23122906/19897317/fc0bf39e-a02d-11e6-89d1-8f44adba2142.png)

**Comment by matlabbe** [comment]
There is a bug in the release version (it is fixed in source) in which MLS is still running even if Organized Cloud reconstruction is selected. To fix this, select dense cloud reconstruction flavor, this will make visible the MLS group box, then uncheck it. Select back organized cloud reconstruction flavor.

**Comment by Lhun** [comment]
Okay, so, success! That was indeed the issue - For anyone else experiencing the same - I did the following: 

Uninstall of rtab map using the provided uninstaller. Cleared .rtabmap in the users directory and dumped my documents/rtab-map directory. 

Uninstalled openNi2, reinstalled the latest from their website. 
Re-installed Rtabmap 0.11.11. Imported the included database, saved the optimizer strategy changes and exported, successfully, to an obj. 

I then followed this: 
"There is a bug in the release version (it is fixed in source) in which MLS is still running even if Organized Cloud reconstruction is selected. To fix this, select dense cloud reconstruction flavor, this will make visible the MLS group box, then uncheck it. Select back organized cloud reconstruction flavor." 

and 
re-organizing now also seems to work, as well as decimation. I'm testing cloud filtering as well, but so far things seem great.

Thanks again to @matlabbe , I would say it's safe to say this is no longer an issue.  Well done, I can't wait for the fixes to dense!
